### PR TITLE
feat: Add support for CAPX

### PR DESCRIPTION
**Pull request by deepakm-ntnx**
to regenerate example Cluster with topology yamls
<pre>
 make examples.sync
</pre>

To run unit tests
<pre>
 make test
</pre>

To deploy cre controller as well as nutanix provider and cluster class of nutanix
<pre>
make dev.run-on-kind 
</pre>

To deploy nutanix cluster with topology with cre annotations 
<pre>
export KUBECONFIG=/go/src/github.com/deepakm-ntnx/capi-runtime-extensions/.local/kind/capi-runtime-extensions-dev/kubeconfig
clusterctl generate cluster nutanix-quick-start-helm-addon-cilium \
  --from examples/capi-quick-start/nutanix-cluster-cilium-helm-addon.yaml \
  --kubernetes-version v1.29.1 \
  --worker-machine-count 1 | \
  kubectl apply --server-side -f -

New clusterctl version available: v1.6.1 -> v1.6.2
sigs.k8s.io/cluster-api
Warning: Cluster refers to ClusterClass test-cre in the topology but it does not exist. Cluster topology has not been fully validated. The ClusterClass must be created to reconcile the Cluster
cluster.cluster.x-k8s.io/nutanix-quick-start-helm-addon-cilium serverside-applied
</pre>

### CODE REVIEW

File: api/v1alpha1/constants.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2023 D2iQ, Inc. All rights reserved.
+// Copyright 2024 D2iQ, Inc. All rights reserved.
```
Comment (thunderboltsid) [inline api/v1alpha1/constants.go:1]: nit: would advise running this by legal first.

File: charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 2]
```diff
@@ -0,0 +1,708 @@
...
+            kubeletExtraArgs:
+              cloud-provider: external
+              eviction-hard: nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<15%,memory.available<100Mi,imagefs.inodesFree<10%
+              tls-cipher-suites: ${TLS_CIPHER_SUITES=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256}
+        postKubeadmCommands:
+          - echo export KUBECONFIG=/etc/kubernetes/admin.conf >> /root/.bashrc
+          - echo "after kubeadm call" > /var/log/postkubeadm.log
+        preKubeadmCommands:
+          - echo "before kubeadm call" > /var/log/prekubeadm.log
+          - hostnamectl set-hostname "{{ ds.meta_data.hostname }}"
```
Comment (thunderboltsid) [inline charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml:642]: This might have to be modified to accommodate 1.29 https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/382

File: charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 3]
```diff
@@ -0,0 +1,708 @@
...
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: NutanixMachineTemplate
+metadata:
+  labels:
+    cluster.x-k8s.io/provider: infrastructure-nutanix
+  name: ${CLUSTER_CLASS_NAME}-md-nmt
+  namespace: ${NAMESPACE}
+spec:
+  template:
+    spec:
```
Comment (thunderboltsid) [inline charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml:693]: We should ideally also add:
- project
- additionalCategories
- opus

File: api/v1alpha1/nutanix_clusterconfig_types.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			Description: "Nutanix cluster configuration",
+			Type:        "object",
+			Properties: map[string]clusterv1.JSONSchemaProps{
+				"controlPlaneEndpoint": NutanixControlPlaneEndpointSpec{}.VariableSchema().OpenAPIV3Schema,
+			},
+		},
+	}
+}
+
+type NutanixControlPlaneEndpointSpec struct {
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply api/v1alpha1/nutanix_clusterconfig_types.go:26]: still trying to figure out if it makes sense to be under spec/template/spec/topology/clusterConfig//nutanix/controlPlane/{host, port} or under spec/template/spec/topology/clusterConfig//nutanix/controlPlaneEndpoint/{host, port}

File: api/v1alpha1/constants.go (hunk 1 above)
Comment (deepakm-ntnx) [reply api/v1alpha1/constants.go:1]: @jimmidyson is planning to change it later. changing year is a standard practice.

File: charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 3 above)
Comment (deepakm-ntnx) [reply charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml:693]: yes we need to figure out how to add required vs optional fields so that user can choose either way. this can be added in subsequent PRs

File: charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml [hunk 5]
```diff
@@ -0,0 +1,708 @@
...
+    - kind: ConfigMap
+      name: ${CLUSTER_NAME}-pc-trusted-ca-bundle
+  strategy: ApplyOnce
+---
+apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+kind: KubeadmConfigTemplate
+metadata:
+  labels:
+    cluster.x-k8s.io/provider: infrastructure-nutanix
+  name: ${CLUSTER_NAME}-kcfg-0
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml:278]: this is autogenerated, lets fix it in the main code if needed

File: charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 2 above)
Comment (deepakm-ntnx) [reply charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml:642]: currently we are using alpha1 version. if we cut alpha2, then latest changes will show up

File: hack/examples/bases/nutanix/calico/crs/kustomization.yaml.tmpl [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2024 D2iQ, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+resources:
+- https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/releases/download/v1.4.0-alpha.1/cluster-template-topology.yaml
+
+patches:
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply hack/examples/bases/nutanix/calico/crs/kustomization.yaml.tmpl:10]: resolved by adding $$var instead of $var in hack/examples/bases/nutanix/kustomization.yaml.tmpl for all the env var ref.

File: api/v1alpha1/clusterconfig_types.go [hunk 7]
```diff
@@ -40,6 +40,8 @@ type ClusterConfigSpec struct {
 	AWS *AWSSpec `json:"aws,omitempty"`
 	// +optional
 	Docker *DockerSpec `json:"docker,omitempty"`
+	// +optional
+	Nutanix *NutanixSpec `json:"nutanix,omitempty"`
```
Comment (dkoshkin) [inline api/v1alpha1/clusterconfig_types.go:44]: To make the reviewing easier what do you think about opening a PR with just API changes for only the required input?

File: api/v1alpha1/nutanix_node_types.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+	// Defines the boot type of the virtual machine. Only supports UEFI and Legacy
+	BootType NutanixBootType `json:"bootType,omitempty"`
+
+	// systemDiskSize is size (in Quantity format) of the system disk of the VM
+	// The minimum systemDiskSize is 20Gi bytes
+	SystemDiskSize string `json:"systemDiskSize"`
+
+	// List of GPU devices that need to be added to the machines.
+	GPUs []NutanixGPU `json:"gpus,omitempty"`
```
Comment (dkoshkin) [inline api/v1alpha1/nutanix_node_types.go:100]: Maybe the GPU API should be added later after testing just a basic cluster?

File: charts/capi-runtime-extensions/templates/clusterclass_nutanix.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+# Copyright 2023 D2iQ, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+{{- if and .Values.deployDefaultClusterClasses (.Capabilities.APIVersions.Has "infrastructure.cluster.x-k8s.io/v1beta1/NutanixClusterTemplate") }}
```
Comment (dkoshkin) [inline charts/capi-runtime-extensions/templates/clusterclass_nutanix.yaml:4]: Maybe the changes to pull in the ClusterClass can be its own PR?

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml [hunk 10]
```diff
@@ -0,0 +1,117 @@
+apiVersion: cluster.x-k8s.io/v1beta1
```
Comment (dkoshkin) [inline examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml:1]: Please remove `examples/capi-quick-start/aws-cluster-identity.yaml` and `examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml`.

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                    value: value2
+                bootType: legacy
+                cluster:
+                  name: ${NUTANIX_PRISM_ELEMENT_CLUSTER_NAME}
+                  type: name
+                gpus:
+                  - name: gpu1
+                    type: name
+                  - name: gpu2
+                    type: name
```
Comment (dkoshkin) [inline examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml:46]: Lets have just a basic example that will work without special hardware?

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                image:
+                  name: ${NUTANIX_MACHINE_TEMPLATE_IMAGE_NAME}
+                  type: name
+                memorySize: 4Gi
+                project:
+                  name: prj1
```
Comment (dkoshkin) [inline examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml:52]: We want this to be a variable

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+              host: ${CONTROL_PLANE_ENDPOINT_IP}
+              port: ${CONTROL_PLANE_ENDPOINT_PORT}
+            failureDomains:
+              - cluster:
+                  name: PE1
+                  type: name
+                controlPlane: false
+                name: fd1
+                subnets:
+                  - name: subnet1
```
Comment (dkoshkin) [inline examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml:71]: Should be a variable

File: hack/examples/bases/nutanix/calico/crs/kustomization.yaml.tmpl [hunk 14]
```diff
@@ -0,0 +1,27 @@
+# Copyright 2024 D2iQ, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
```
Comment (dkoshkin) [inline hack/examples/bases/nutanix/calico/crs/kustomization.yaml.tmpl:5]: Lets start with only the requirement which is Cilium and then we can add Calico support in the future.
(I'm trying to be mindful of the size of the PR and the test matrix)

File: hack/examples/bases/nutanix/cilium/crs/kustomization.yaml.tmpl [hunk 15]
```diff
@@ -0,0 +1,28 @@
+# Copyright 2024 D2iQ, Inc. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
```
Comment (dkoshkin) [inline hack/examples/bases/nutanix/cilium/crs/kustomization.yaml.tmpl:2]: I think this would be a good separate PR, just the changes needed to create example files and create the ClusterClass in the bootstrap. Without any of the API changes or handlers.

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml (hunk 12 above)
Comment (deepakm-ntnx) [reply examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml:52]: already tracked internally on capx

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml (hunk 13 above)
Comment (deepakm-ntnx) [reply examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml:71]: already tracked on capx

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml (hunk 10 above)
Comment (deepakm-ntnx) [reply examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml:288]: removed aws-cluster-identity.yaml but nutanix-cluster-calico-helm-addon.yaml is now up to date

File: examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml (hunk 11 above)
Comment (deepakm-ntnx) [reply examples/capi-quick-start/nutanix-cluster-calico-helm-addon.yaml]: removed.

**Comment by jimmidyson** [comment]
I noticed conflicts so I rebased this and fixed up (force pushed) @deepakm-ntnx. Please pull latest changes locally.

**Comment by jimmidyson** [comment]
Fixed up conflicts again after repo rename.

File: charts/capi-runtime-extensions/templates/clusterclass_nutanix.yaml (hunk 9 above)
Comment (deepakm-ntnx) [reply charts/capi-runtime-extensions/templates/clusterclass_nutanix.yaml]: separated thru https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/473

File: hack/examples/bases/nutanix/calico/crs/kustomization.yaml.tmpl (hunk 14 above)
Comment (deepakm-ntnx) [reply hack/examples/bases/nutanix/calico/crs/kustomization.yaml.tmpl]: we will have to modify sync.sh to ignore non existent folders. keeping it short thru https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/473

File: hack/examples/bases/nutanix/cilium/crs/kustomization.yaml.tmpl (hunk 15 above)
Comment (deepakm-ntnx) [reply hack/examples/bases/nutanix/cilium/crs/kustomization.yaml.tmpl]: https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/473

**Comment by deepakm-ntnx** [comment]
closing as due to rename moved to https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/6/files

File: api/v1alpha1/nutanix_clusterconfig_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply api/v1alpha1/nutanix_clusterconfig_types.go:69]: https://d2iq.atlassian.net/browse/D2IQ-100365

File: charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml (hunk 2 above)
Comment (deepakm-ntnx) [reply charts/capi-runtime-extensions/defaultclusterclasses/nutanix-cluster-class.yaml]: handled in https://github.com/d2iq-labs/cluster-api-runtime-extensions-nutanix/pull/6
